connectors/utils.py

```diff
@@ -422,52 +419,51 @@ def _callback(self, task, result_callback=None):
         elif task.exception():
             logger.error(
                 f"Exception found for task {task.get_name()}: {task.exception()}",
-                exc_info=True,
             )
-        if result_callback is not None:
-            result_callback(task.result())
-        # global callback
-        if self.results_callback is not None:
-            self.results_callback(task.result())
 
-    def _add_task(self, coroutine, result_callback=None):
+    def _add_task(self, coroutine):
         task = asyncio.create_task(coroutine())
         self.tasks.append(task)
         # _callback will be executed when the task is done,
         # i.e. the wrapped coroutine either returned a value, raised an exception, or the Task was cancelled.
         # Ref: https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.done
         task.add_done_callback(
-            functools.partial(self._callback, result_callback=result_callback)
+            functools.partial(self._callback)
         )
         return task
 
-    async def put(self, coroutine, result_callback=None):
+    async def put(self, coroutine):
         """Adds a coroutine for immediate execution.
 
         If the number of running tasks reach `max_concurrency`, this
         function will block and wait for a free slot.
-
-        If provided, `result_callback` will be called when the task is done.
         """
         await self._sem.acquire()
-        return self._add_task(coroutine, result_callback=result_callback)
+        return self._add_task(coroutine)
 
-    def try_put(self, coroutine, result_callback=None):
+    def try_put(self, coroutine):
         """Tries to add a coroutine for immediate execution.
 
         If the number of running tasks reach `max_concurrency`, this
         function return a None task immediately
-
-        If provided, `result_callback` will be called when the task is done.
         """
 
         if self._sem.try_acquire():
-            return self._add_task(coroutine, result_callback=result_callback)
+            return self._add_task(coroutine)
         return None
 
-    async def join(self):
+    async def join(self, raise_on_error=False):
         """Wait for all tasks to finish."""
-        await asyncio.gather(*self.tasks, return_exceptions=True)
+        await asyncio.gather(*self.tasks, return_exceptions=(not raise_on_error))
+
+    def raise_any_exception(self):
+        for task in self.tasks:
+            if task.done() and not task.cancelled():
+                if task.exception():
+                    logger.error(
+                        f"Exception found for task {task.get_name()}: {task.exception()}",
+                    )
+                    raise task.exception()
```

Comment:
It looks good, but also thinking out loud - what if there are multiple errors?

We're _probably_ okay raising only the first one.

Alternatives:
- Collect and log all errors
- Raise the first one (mentioning that we're raising the first one)
- Wrap all errors in one big errors

I think we can just roll with what we have here and address this later

Reply:
my _guess_ if we have multiple errors is that they'll all be the same. And since they should already be getting logged by the default callback defined in the ConcurrentTasks abstraction, I think we're also ok with this as is. But I agree that this would be a good spot to come back to if/when we have something like the WS Monitor that can hold multiple errors in state for a diagnostic bundle.